File: src/extras-kio/libpaste.cpp
#include <KFileItem>
#include <QMimeData>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <paste.h>
#include "libpaste.h"
#include "libpaste.hxx"

bool KIO_CanPasteMimeData(const QMimeData* param1) {
    return KIO::canPasteMimeData(param1);
}

libqt_string KIO_PasteActionText(const QMimeData* param1, bool* param2, const KFileItem* param3) {
    QString _ret = KIO::pasteActionText(param1, param2, *param3);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void KIO_SetClipboardDataCut(QMimeData* param1, bool param2) {
    KIO::setClipboardDataCut(param1, param2);
}

bool KIO_IsClipboardDataCut(const QMimeData* param1) {
    return KIO::isClipboardDataCut(param1);
}

File: src_original/extras-kio/libpaste.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kio.html
pub const kio = struct {
    /// [Qt documentation](https://api.kde.org/kio.html#canPasteMimeData)
    ///
    /// ``` param1: QtC.QMimeData ```
    pub fn CanPasteMimeData(param1: ?*anyopaque) bool {
        return qtc.KIO_CanPasteMimeData(@ptrCast(param1));
    }

    /// [Qt documentation](https://api.kde.org/kio.html#pasteActionText)
    ///
    /// ``` param1: QtC.QMimeData, param2: *bool, param3: QtC.KFileItem, allocator: std.mem.Allocator ```
    pub fn PasteActionText(param1: ?*anyopaque, param2: *bool, param3: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KIO_PasteActionText(@ptrCast(param1), @ptrCast(param2), @ptrCast(param3));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio.PasteActionText: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio.html#setClipboardDataCut)
    ///
    /// ``` param1: QtC.QMimeData, param2: bool ```
    pub fn SetClipboardDataCut(param1: ?*anyopaque, param2: bool) void {
        qtc.KIO_SetClipboardDataCut(@ptrCast(param1), param2);
    }

    /// [Qt documentation](https://api.kde.org/kio.html#isClipboardDataCut)
    ///
    /// ``` param1: QtC.QMimeData ```
    pub fn IsClipboardDataCut(param1: ?*anyopaque) bool {
        return qtc.KIO_IsClipboardDataCut(@ptrCast(param1));
    }
};
